### PR TITLE
nonon

### PR DESCRIPTION
**Pull request by jdgc**

### CODE REVIEW

File: contracts/Nonon/NononFriendCard.sol [hunk 1]
```diff
@@ -0,0 +1,336 @@
...
+            if (to != from) {
+                sentBitmap[from].setBatch(collectionTokenStartId, quantity);
+            }
+        }
+
+        if (to != address(0)) {
+            receivedBitmap[to].setBatch(collectionTokenStartId, quantity);
+        }
+
+        emit BatchMetadataUpdate(1, type(uint256).max);
```
Comment (kneelsdev) [inline contracts/Nonon/NononFriendCard.sol:267]: What is the use for this event?

File: contracts/Nonon/NononFriendCard.sol [hunk 2]
```diff
@@ -0,0 +1,336 @@
...
+        return receivedBitmap[owner].get(tokenId);
+    }
+
+    // check if given address has ever sent tokenId
+    function hasSentToken(address owner, uint256 tokenId) external view returns (bool) {
+        return sentBitmap[owner].get(tokenId);
+    }
+
+    function tokenStatusMap(address owner, bool sent) external view returns (uint256[] memory received) {
+        // TODO: reference max supply instead of hardcoding
```
Comment (kneelsdev) [inline contracts/Nonon/NononFriendCard.sol:317]: Reminder for this todo

File: contracts/Nonon/NononFriendCard.sol [hunk 3]
```diff
@@ -0,0 +1,336 @@
...
+    function points(uint256 tokenId) public view returns (uint256) {
+        address owner = ownerOf(tokenId);
+        uint256 max = IERC721A(collectionAddress).totalSupply() + 1;
+
+        return receivedBitmap[owner].popCount(1, max) + sentBitmap[owner].popCount(1, max);
+    }
+
+    // convenience function to get point information in a token range
+    // note that this is expensive and most likely will require multiple calls to cover large ranges.
+    function tokenPointsInRange(uint256 startId, uint256 endId) public view returns (TokenPoints[] memory) {
```
Comment (kneelsdev) [inline contracts/Nonon/NononFriendCard.sol:280]: Is this used internally? or should it be an external method?

File: contracts/Nonon/NononFriendCard.sol [hunk 4]
```diff
@@ -0,0 +1,337 @@
...
+    function points(uint256 tokenId) public view returns (uint256) {
+        address owner = ownerOf(tokenId);
+        uint256 max = IERC721A(collectionAddress).totalSupply() + 1;
+
+        return receivedBitmap[owner].popCount(1, max) + sentBitmap[owner].popCount(1, max);
+    }
+
+    // convenience function to get point information in a token range
+    // note that this is expensive and most likely will require multiple calls to cover large ranges.
+    function tokenPointsInRange(uint256 startId, uint256 endId) external view returns (TokenPoints[] memory) {
```
Comment (jdgc) [inline contracts/Nonon/NononFriendCard.sol:282]: changed this to external

File: contracts/Nonon/NononFriendCard.sol [hunk 5]
```diff
@@ -0,0 +1,337 @@
...
+        return receivedBitmap[owner].get(tokenId);
+    }
+
+    // check if given address has ever sent tokenId
+    function hasSentToken(address owner, uint256 tokenId) external view returns (bool) {
+        return sentBitmap[owner].get(tokenId);
+    }
+
+    function tokenStatusMap(address owner, bool sent) external view returns (uint256[] memory received) {
+        uint256 maxWordIndex = NONON_MAX_SUPPLY >> 8;
```
Comment (jdgc) [inline contracts/Nonon/NononFriendCard.sol:319]: decided to use a constant here in this specific case for simplicity

File: contracts/Nonon/NononFriendCard.sol (hunk 3 above)
Comment (jdgc) [reply contracts/Nonon/NononFriendCard.sol]: changed to external in https://github.com/guminc/contracts/pull/73/commits/67285f429a95c1766f6f0e11e1d2c44eb1eeb991

File: contracts/Nonon/NononFriendCard.sol (hunk 2 above)
Comment (jdgc) [reply contracts/Nonon/NononFriendCard.sol]: went with a constant in this case, added it and removed comment in https://github.com/guminc/contracts/pull/73/commits/67285f429a95c1766f6f0e11e1d2c44eb1eeb991

**Pull request closed without merging**